File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.py
import sys
def bsearch(M,highway):
    localmin=1
    localmax=0
    answer=0
    for i in range(1,N+2):
        localmax=max(localmax,highway[i]-highway[i-1])
    while(localmin<=localmax):
        localmid = (localmax+localmin)//2
        cnt=0
        for i in range(1,N+2):
            cnt += (highway[i] - highway[i - 1]) // localmid - 1
            if ((highway[i] - highway[i - 1]) % localmid>0):
                cnt+=1
        if(cnt<=M):
            answer=localmid
            localmax=localmid-1
        else:
            localmin=localmid+1
    return answer
input,print = sys.stdin.readline,sys.stdout.write
N,M,L=map(int,input().split(" "))
if(N>0):
    highway=list(map(int,input().split(" ")))
    highway.append(0)
    highway.append(L)
    highway.sort()
else:
    highway=[0,L]
answer=bsearch(M,highway)
print(f"{answer}")

File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.cc
#include <iostream>
#include <algorithm>
using namespace std;
int N, M, L,highway[52],localmax=0,localmin=1,answer;
void bsearch() {
	while (localmin <= localmax) {
		int localmid = (localmax + localmin) / 2;
		int cnt = 0;
		for (int i = 1; i <= N + 1; i++) {
			cnt += (highway[i] - highway[i - 1]) / localmid - 1;
			if ((highway[i] - highway[i - 1]) % localmid)cnt++;
		}
		if (cnt <= M) {
			answer = localmid;
			localmax = localmid - 1;
		}
		else localmin = localmid + 1;
	}
}
int main(void) {
	ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
	cin >> N >> M >> L;
	for (int i = 0; i < N; i++)cin >> highway[i];
	highway[N + 1] = L;
	sort(highway, highway + N+2);
	for (int i = 1; i <= N + 1; i++) {
		localmax = max(localmax, highway[i] - highway[i - 1]);
	}
	bsearch();
	cout << answer;
	return 0;
}